qiskit/aqua/operators/primitive_ops/pauli_sum_op.py

```diff
@@ -0,0 +1,378 @@
+                        ) + new_dict.get(n_str, 0)
+                    return DictStateFn(new_dict, coeff=self.coeff * front.coeff)
+
+            elif isinstance(front, StateFn) and front.is_measurement:
+                raise ValueError("Operator composed with a measurement is undefined.")
+
+            # Composable types with PauliOp
+            elif isinstance(front, (PauliSumOp, PauliOp, CircuitOp, CircuitStateFn)):
+                return self.compose(front).eval()  # type: ignore
+
+        # Covers VectorStateFn and OperatorStateFn
+        return self.to_matrix_op().eval(front.to_matrix_op())  # type: ignore
+
+    def exp_i(self) -> OperatorBase:
+        """ Return a ``CircuitOp`` equivalent to e^-iH for this operator H. """
+        # TODO: optimize for some special cases
+        from ..evolutions.evolved_op import EvolvedOp
+
+        return EvolvedOp(self)
+
+    def to_instruction(self) -> Instruction:
+        return self.to_matrix_op().to_circuit().to_instruction()  # type: ignore
+
+    def to_pauli_op(self, massive: bool = False) -> OperatorBase:
+        from .pauli_op import PauliOp
+
+        def to_real(x):
+            return x.real if np.isreal(x) else x
+
+        def to_native(x):
+            return x.item() if isinstance(x, np.generic) else x
+
+        if len(self.primitive) == 1:
+            return PauliOp(
+                Pauli(x=self.primitive.table.X[0], z=self.primitive.table.Z[0]),  # type: ignore
+                to_native(to_real(self.primitive.coeffs[0])) * self.coeff,  # type: ignore
+            )
+        return SummedOp(
+            [
+                PauliOp(
+                    Pauli(x=s.table.X[0], z=s.table.Z[0]),
+                    to_native(to_real(s.coeffs[0])),
+                )
+                for s in self.primitive
+            ],
+            coeff=self.coeff,
+        )
+
+    def __getitem__(self, offset: Union[int, slice]) -> "PauliSumOp":
+        """Allows array-indexing style access to the ``PauliSumOp``.
+
+        Args:
+            offset: The index of ``PauliSumOp``.
+
+        Returns:
+            The ``PauliSumOp`` at index ``offset``,
+        """
+        return PauliSumOp(self.primitive[offset], self.coeff)
+
+    def __len__(self) -> int:
+        """Length of ``SparsePauliOp``.
+
+        Returns:
+            An int equal to the length of SparsePauliOp.
+        """
+        return len(self.primitive)
+
+    # pylint: disable=arguments-differ
+    def reduce(self, atol: Optional[float] = None, rtol: Optional[float] = None) -> "PauliSumOp":
+        """Simplify the primitive ``SparsePauliOp``.
+
+        Args:
+            atol: Absolute tolerance for checking if coefficients are zero (Default: 1e-8).
+            rtol: Relative tolerance for checking if coefficients are zero (Default: 1e-5).
+
+        Returns:
+            The simplified ``PauliSumOp``.
+        """
+        if isinstance(self.coeff, Number):
```

Comment:
Similar comment to here about instance check could check for not a ParameterExpression?